src/main/capi/arrow-c.cpp

```diff
@@ -108,3 +108,112 @@ duckdb_state duckdb_execute_prepared_arrow(duckdb_prepared_statement prepared_st
 	*out_result = (duckdb_arrow)arrow_wrapper;
 	return !arrow_wrapper->result->HasError() ? DuckDBSuccess : DuckDBError;
 }
+
+namespace arrow_array_stream_wrapper {
+namespace {
+struct PrivateData {
+	ArrowSchema *schema;
+	ArrowArray *array;
+	bool done;
+};
+
+void EmptySchemaRelease(ArrowSchema *schema) {
+}
+
+void GetSchemaVoid(uintptr_t stream_factory_ptr, duckdb::ArrowSchemaWrapper &schema) {
+	auto private_data = (PrivateData *)((ArrowArrayStream *)stream_factory_ptr)->private_data;
+	schema.arrow_schema = *private_data->schema;
+	schema.arrow_schema.release = EmptySchemaRelease;
+
+	for (int64_t i = 0; i < schema.arrow_schema.n_children; i++) {
+		schema.arrow_schema.children[i]->release = EmptySchemaRelease;
+	}
+}
+
+int GetSchema(struct ArrowArrayStream *stream, struct ArrowSchema *out) {
+	auto private_data = static_cast<arrow_array_stream_wrapper::PrivateData *>((stream->private_data));
+	*out = *private_data->schema;
+	return 0;
+}
+
+int GetNext(struct ArrowArrayStream *stream, struct ArrowArray *out) {
+	auto private_data = static_cast<arrow_array_stream_wrapper::PrivateData *>((stream->private_data));
+	*out = *private_data->array;
+	if (private_data->done) {
+		out->release(out);
+	}
+
+	private_data->done = true;
+	return 0;
+}
+
+const char *GetLastError(struct ArrowArrayStream *stream) {
+	return nullptr;
+}
+
+void Release(struct ArrowArrayStream *) {
+}
+
+duckdb_state Ingest(duckdb_connection connection, const char *table_name, struct ArrowArrayStream *input) {
+	try {
+		auto cconn = (duckdb::Connection *)connection;
+		cconn
+		    ->TableFunction("arrow_scan", {duckdb::Value::POINTER((uintptr_t)input),
+		                                   duckdb::Value::POINTER((uintptr_t)input->get_next),
+		                                   duckdb::Value::POINTER((uintptr_t)GetSchemaVoid)})
+		    ->CreateView(table_name, true, true);
+	} catch (const std::exception &exc) {
+		// TODO: We should probably log this, but where?
+		return DuckDBError;
+	}
+
+	return DuckDBSuccess;
+}
+} // namespace
+} // namespace arrow_array_stream_wrapper
+
+duckdb_state duckdb_arrow_scan(duckdb_connection connection, const char *table_name, duckdb_arrow_stream arrow) {
+	auto stream = reinterpret_cast<ArrowArrayStream *>(arrow->__arrwstr);
+
+	// Backup release functions - we nullify children schema release functions because we don't want to release on
+	// behalf of the caller, downstream in our code. Note that Arrow releases target immediate children, but aren't
+	// recursive. So we only back up immediate children here and restore their functions.
```

Comment:
I'm not yet sure about this. If I understand it correctly, the idea is that Arrow is then not released by the DuckDB Scan?

Reply:
I see, thanks for the clarification :-)